### PR TITLE
cosmo: make derived_parameters private

### PR DESCRIPTION
**Pull request by nstarman**
- [ ] What's New entry directly in v6.0.x after merging this and backporting it to v6.0.x

I was working on a followup to #15484 (for v6.1) and realized there was a conceptual issue.

It’s unclear why `Ode0` should be “derived” but not `Ok0` or `Onu0` except that in the superclass it’s a Parameter object. It’s important for cosmology methods, eg. equality, to know this, but not the user. In fact it's confusing.

Therefore I think the new ``derived_parameters`` object should be made private. ``cosmology_class.Ode0`` still works so there's no API change yet.

As an aside I think the better solution is to have a "descriptors" (name not final) attribute that just collects all the descriptors that will be hidden from the user as proper dataclass descriptor objects are. That's for a future PR.

Sorry @eerovaher for the change right after #15168.

### CODE REVIEW

File: docs/whatsnew/6.0.rst [hunk 1]
```diff
@@ -210,8 +208,6 @@ with information about the parameter used when setting up the cosmology::
     >>> from astropy.cosmology import FlatLambdaCDM
     >>> FlatLambdaCDM.parameters["H0"]
     Parameter(derived=False, unit=Unit("km / (Mpc s)"), equivalencies=[], ...)
-    >>> FlatLambdaCDM.derived_parameters["Ode0"]
-    Parameter(derived=True, unit=None, equivalencies=[], ...)
```
Comment (eerovaher) [inline docs/whatsnew/6.0.rst:214]: It might be better to keep the example with `"Ode0"` and to delete the one with `"H0"` so that this code block and the preceding one would access different parameters.

File: docs/whatsnew/6.0.rst [hunk 2]
```diff
@@ -191,17 +191,15 @@ format::
 New properties to access |Cosmology| parameters
 -----------------------------------------------
 
-The :class:`~astropy.cosmology.Cosmology` class now has two new properties
-to access the parameters of the cosmology: :attr:`~astropy.cosmology.Cosmology.parameters`
-and :attr:`~astropy.cosmology.Cosmology.derived_parameters`. These properties
-return a :class:`~types.MappingProxyType` object, which is a read-only dictionary
-of all the (derived-)parameter values on the |Cosmology| instance. For example::
+The :class:`~astropy.cosmology.Cosmology` class now has a new property to access the
```
Comment (eerovaher) [inline docs/whatsnew/6.0.rst:194]: If you are editing this anyways you might as well shorten it:
```suggestion
The |Cosmology| class now has a new property to access the
```

File: docs/changes/cosmology/15168.feature.rst [hunk 3]
```diff
@@ -1,5 +1,4 @@
-The ``Cosmology`` class now has two new properties to access the parameters of the
-cosmology: ``.parameters`` and ``.derived_parameters``. These properties return a
-read-only dictionary of all the (derived) parameter values on the cosmology object.
-When accessed from the class (not an instance) the dictionary contains ``Parameter``
-objects, not the values.
+The ``Cosmology`` class now has a new property to access the parameters of the
+cosmology: ``.parameters``. This property return a read-only dictionary of all the
+(derived) parameter values on the cosmology object. When accessed from the class (not an
+instance) the dictionary contains ``Parameter`` objects, not the values.
```
Comment (eerovaher) [inline docs/changes/cosmology/15168.feature.rst:4]: Is this better phrasing?
```suggestion
instance) the dictionary contains ``Parameter`` instances, not the values.
```

File: docs/whatsnew/6.0.rst (hunk 1 above)
Comment (nstarman) [reply docs/whatsnew/6.0.rst:214]: Can you elaborate? ``Ode0`` is in `_derived_parameters`` which is private.

**Comment by nstarman** [comment]
@pllim this PR edits a changelog, so the CI is unhappy about that.

RTD failure is unrelated?

**Comment by mhvk** [comment]
Ideally, this PR does go in 6.0 to avoid adding/changing API twice. Note that I'm not sure this is all that necessary: it is fine to tell the user which parameters are counted as derived even if in principle other choices might be possible.

**Comment by nstarman** [comment]
> Ideally, this PR does go in 6.0 to avoid adding/changing API twice.

Agreed. 

> it is fine to tell the user which parameters are counted as derived even if in principle other choices might be possible.

There are two things on this.

1. I want to make some pedagogically useful cosmology classes, like 2-component universes with just matter & radiation (early universe) or just matter & dark energy (later universe). The choices of parameters is definitional. The differences between all the classes' ``derived_parameters`` is not well motivated. I don't think users need this window into the factory where the sausage is made.

2. I want to speed up / simplify the construction of FLRW. Currently all these extra properties like ``hubble_distance`` are computed in ``__init__``. Combining a frozen dataclass (like Cosmology will be) and a non-data descriptor (one that has a ``__get__`` but not ``__set__`` or ``__delete__``) it's easy to make a lazy property that caches in ``__dict__`` and is still immutable. Therefore we can make all the stuff in ``FLRW.__init__`` be both cached and lazily computed (it's currently just cached), speeding up every cosmology instance. I mention this because these are all "derived" properties. I don't want to provide mixed messaging to users.

**Comment by nstarman** [comment]
@saimn, what should I do about the whatsnew conflict?

**Comment by mhvk** [comment]
@nstarman - reasoning certainly makes sense. And much easier to make something public later than making it private after it has been public.

File: astropy/cosmology/core.py [hunk 4]
```diff
@@ -96,15 +96,15 @@ class Cosmology(metaclass=abc.ABCMeta):
     write = UnifiedReadWriteMethod(CosmologyWrite)
 
     # Parameters
-    parameters = ParametersAttribute()
+    parameters = ParametersAttribute(attr_name="_parameters")
```
Comment (mhvk) [inline astropy/cosmology/core.py:99]: Don't really have time for a full review, but why the new `attr_name`? It seems unrelated to the PR and not really used.

Comment (nstarman) [reply astropy/cosmology/core.py:99]: It's because the current form assumes the ``ParametersAttribute`` is connected to an underscore-prefixed class attribute that's created in `Cosmology.__init_subclass__`. For a private attribute that assumption means the connected attribute is double-underscore-prefixed! This is of course a problem because attribute access is different for double-underscore-prefixed names and ``getattr(instance, "__derived_paramters")`` doesn't work. 

This PR changes ``__set_name__` from building ``attr_name`` based upon this problematic assumption to having ``attr_name`` be set at ``ParametersAttribute`` construction so as to avoid double-underscore-prefixed attributes. In the next line (100) you'll see that `_derived_parameters` is now connected to `_parameters_derived` (not `__derived_parameters`), which is the dict of derived `Parameter` objects and the counterpart to `_parameters` in this line.

Comment (mhvk) [reply astropy/cosmology/core.py:99]: Ah, fair enough! Though why not use the same trick as `lazyproperty` and just store the actual data under the same name in `parent.__dict__`?

Comment (nstarman) [reply astropy/cosmology/core.py:99]: Does that work for class attributes?

Comment (nstarman) [reply astropy/cosmology/core.py:99]: Unfortunately this doesn't appear to be viable

CleanShot 2023-10-29 at 17 52 57@2x

Comment (nstarman) [reply astropy/cosmology/core.py:99]: The only relevant part of the above image is that `__init_subclass__` can add attributes, but not modify the ``__dict__`` since this is a class-level attribute.

Comment (nstarman) [reply astropy/cosmology/core.py:99]: Unfortunately this doesn't appear to be viable because this is a class variable not an instance variable, like `lazyproperty` works with.

CleanShot 2023-10-29 at 18 00 59@2x

Comment (mhvk) [reply astropy/cosmology/core.py:99]: No, indeed, that would work only on instances... Sorry for the noise!

Comment (nstarman) [reply astropy/cosmology/core.py:99]: I think it would work with a metaclass, but let's avoid that  😆

Comment (nstarman) [reply astropy/cosmology/core.py:99]: I don't love the parameter name `attr_name`. Any suggestions? Maybe `get_attr`...

**Comment by pllim** [comment]
You need to rebase and not skip the CI if you want to push this forward for real. Though you missed the feature freeze, so can we re-milestone this?

File: docs/whatsnew/6.0.rst (hunk 1 above)
Comment (eerovaher) [reply docs/whatsnew/6.0.rst:214]: The change log entry in the version I reviewed said that derived parameters should be accessible through `.parameters`, but the updated version does not state that anymore, so it is clear now that my suggestion is not valid.

**Comment by pllim** [comment]
Hmm that is a bit troublesome, because the What's New page for v6.0 is only in v6.0.x now, not in main anymore.

**Comment by nstarman** [comment]
> Hmm that is a bit troublesome, because the What's New page for v6.0 is only in v6.0.x now, not in main anymore.

Yeah, that was my question to @saimn.

**Comment by saimn** [comment]
The whatsnew change should go only in the `v6.0.x` branch. So maybe rebase and merge on main without the whatnew change, and make another PR on `v6.0.x` with the whatsnew.

**Comment by nstarman** [comment]
@saimn, done!

**Comment by pllim** [comment]
Conflict is gone. Does either @mhvk or @eerovaher want to approve?

**Review by eerovaher — APPROVED** [review]
This pull request could be squashed and merged, but because it does not have its own change log entry and it instead modifies an already existing entry then I don't know how to make the change log checking bot happy.

**Comment by nstarman** [comment]
I think this is where we ask @pllim for help :)

**Comment by pllim** [comment]
With the old change log, we could append PR numbers to existing entry but with this towncrier stuff, I don't know how to do that. I will ask bot to just not check and then squash merge. Thanks, all!

**Comment by pllim** [comment]
The bot is misbehaving. I am just gonna ignore it. Thanks, everyone!
